### ISSUE TITLE
New component: Query Builder

### ISSUE THREAD

**Issue by Rapster** [primefaces-extensions/primefaces-extensions.github.com#432]
I need to know how you feel about implementing this component http://querybuilder.js.org/. I plan to use it in our project, but before making any official annoucement, I want to make sure this component can be "easily" transform into JSF component. I've never implementing JSF component from scratch per say, so I may miss something.

I'm thinking about several iterations. 

- [ ] Reproduce this https://jsfiddle.net/Rapster/89byg4wh/
- [ ] Compatible with jQuery UI CSS Selectors
- [ ] Reuse of PF widgets

Thank you for sharing your thoughts

**Comment by melloware** [comment]
To me this seems like a really useful component especially for reporting type applications where customers want to basically build a query with AND's and OR's etc.

**Comment by ova2** [comment]
I think it is too specific. I already wanted to create such component 5-6 years ago as I was working on DMS with dynamic seach masks. It's up to you guys. I don't use Extensions.

**Comment by Rapster** [comment]
Somehow it feels like a `DynaForm` more stylish and more business specific

**Comment by Rapster** [comment]
I might need your help on this one, so far, I have this:
```xml
<pe:queryBuilder filters="#{queryBuilderController.filters}" locale="fr" />  
```

It displays well, now, I want to do the same think like in this demo http://querybuilder.js.org/demo.html#basic. Meaning, clicking on "Get Rules" will transform the query into JSON. The author of this plugin is able to do so by doing this:
```javascript
var result = $('#builder-basic').queryBuilder('getRules');
```

Basically I want to be able to bind this json result into a property of a bean, and do this:
```xml
<pe:queryBuilder filters="#{queryBuilderController.filters}" locale="fr" value="#{queryBuilderController.value}" />  
```

Any idea how I could do this properly?

**Comment by tandraschko** [comment]
define a behavior event and pass the JSON as request parameter. On server you can catch this at best in #queueEvent.
You have a look at some components, it's a common thing.

**Comment by Rapster** [comment]
Are you saying I'll write something like this:
```html
<pe:queryBuilder filters="#{queryBuilderController.filters}" locale="fr" >
    <p:ajax event="save" actionListener="#{queryBuilderController.saveListener}" />
</pe:queryBuilder>  
```

**Comment by tandraschko** [comment]
Nope.
You can still have a value attribute. But you must define a a behavior event to pass somehow the data from client to server.
Check some of our widgets, we also have something like:
if behaviors.contains['save'] -> reuse, otherwise create a ajax request on the fly.

in queueEvent you can than parse the json from the request parameters and assign it manually to the defined value attribute. Something like:
String json = request.getParamter(component.getClientID() + "_myJson");
MyModel model = parse(json);
component.getValueExpression("value").setValue(model)

**Comment by Rapster** [comment]
I think I see what you mean, but it works only for event right? While basically,  I want to "update" value attribute when form is submitted.

**Comment by tandraschko** [comment]
Then you must store the json in a hidden input. The part in queueEvent should be the same

**Comment by Rapster** [comment]
Does it mean my component should implement something like `EditableValueHolder`, `ValueHolder` or be an `UIInput`? So far, it's just an `UIComponentBase`. I see component like `PickList` inherits from `UIInput`

**Comment by tandraschko** [comment]
1) not required actually. Its just usefull if you need converters and validators
2) you cant. You always need to synchronize your widget and the hidden input on each change.

**Comment by Rapster** [comment]
Thanks for your help @tandraschko I understand now ;) When you said 

> if behaviors.contains['save'] -> reuse, otherwise create a ajax request on the fly.

Does it mean use `PrimeFaces.ajax.AjaxRequest(options);` ? I'm trying to do as `AutoComplete` does _(just to understand how things works)_

**Comment by tandraschko** [comment]
Yeah

**Comment by tandraschko** [comment]
JFYI: if you have some problems to understand something, we can also chat about it somewhere. I like your work on PF/PFE, so just ping me if you need something.

**Comment by Rapster** [comment]
Thanks @tandraschko, so far, I didn't finish my work on this component. We decided to go for a very basic component. IMO this component can be "easily" transform into JSF component, but a lot of work has to be done to cover all features, nd use jQuery UI CSS style. The author of this component is figuring out for a new version, don't know if it would be better idea to wiat for it.

**Comment by Rapster** [comment]
Close for now since my chief and I decided to do otherwise
